src/network/connection_handle.cpp

```diff
@@ -1,188 +1,257 @@
+    // Wait for the connection to become writable.
+    handle->UpdateEventFlags(EV_WRITE | EV_PERSIST);
+    return Transition::NONE;
+  }
+
+  /** Wait for the connection to become readable, or until a timeout happens. */
+  static Transition WaitForReadWithTimeout(const common::ManagedPointer<ConnectionHandle> handle) {
+    handle->UpdateEventFlags(EV_READ | EV_PERSIST | EV_TIMEOUT, READ_TIMEOUT);
+    return Transition::NONE;
+  }
+
+  /** Stop listening to network events. This is used when control is completely ceded to Terrier, hence the name. */
+  static Transition WaitForTerrier(const common::ManagedPointer<ConnectionHandle> handle) {
+    handle->StopReceivingNetworkEvent();
+    return Transition::NONE;
+  }
+};
+
+ConnectionHandle::StateMachine::TransitionResult ConnectionHandle::StateMachine::Delta(ConnState state,
+                                                                                       Transition transition) {
+  // clang-format off
+  switch (state) {
+    case ConnState::READ:      return ConnectionHandleStateMachineTransition::TransitionForRead(transition);
+    case ConnState::PROCESS:   return ConnectionHandleStateMachineTransition::TransitionForProcess(transition);
+    case ConnState::WRITE:     return ConnectionHandleStateMachineTransition::TransitionForWrite(transition);
+    case ConnState::CLOSING:   return ConnectionHandleStateMachineTransition::TransitionForClosing(transition);
+    default:                   throw std::runtime_error("Undefined transition!");
+  }
+  // clang-format on
+}
+
+void ConnectionHandle::StateMachine::Accept(Transition action, const common::ManagedPointer<ConnectionHandle> handle) {
   Transition next = action;
+  // Transition until there are no more transitions.
   while (next != Transition::NONE) {
-    transition_result result = Delta(current_state_, next);
+    TransitionResult result = Delta(current_state_, next);
     current_state_ = result.first;
     try {
-      next = result.second(connection);
-    } catch (NetworkProcessException &e) {
+      next = result.second(handle);
+    } catch (const NetworkProcessException &e) {
+      // If an error occurs, the error is logged and the connection is terminated.
       NETWORK_LOG_ERROR("{0}\n", e.what());
       next = Transition::TERMINATE;
     }
   }
 }
 
+ConnectionHandle::ConnectionHandle(int sock_fd, common::ManagedPointer<ConnectionHandlerTask> task,
+                                   common::ManagedPointer<trafficcop::TrafficCop> tcop,
+                                   std::unique_ptr<ProtocolInterpreter> interpreter)
+    : io_wrapper_(std::make_unique<NetworkIoWrapper>(sock_fd)),
+      conn_handler_task_(task),
+      traffic_cop_(tcop),
+      protocol_interpreter_(std::move(interpreter)) {
+  context_.SetCallback(Callback, this);
+  context_.SetConnectionID(static_cast<connection_id_t>(sock_fd));
+}
+
+ConnectionHandle::~ConnectionHandle() { context_.Reset(); }
```

Comment:
Is reseting necessary if the `ConnectionContext` is about be to destroyed? I took a peek at `Reset` and it doesn't seem necessary to be called here.

Reply:
you're right